### PR TITLE
stop scheduler from advancing through empty buckets without accept from ratelimiter

### PR DESCRIPTION
**Pull request by jim-minter**
I believe this fixes https://bugzilla.redhat.com/show_bug.cgi?id=1515058 and substantially helps https://bugzilla.redhat.com/show_bug.cgi?id=1543446

### CODE REVIEW

**Comment by jim-minter** [comment]
@bparees do you think this need backporting?

**Comment by bparees** [comment]
> return (s.position + inc + len(s.buckets)) % len(s.buckets)

unrelated to your changes but since it hurt my brain, isn't that numerically identical to `return (s.position + inc) % len(s.buckets)`  ?

isn't `(x+y) % y == x % y` ?

File: pkg/image/controller/scheduler.go [hunk 1]
```diff
@@ -69,16 +69,12 @@ func (s *Scheduler) next() (interface{}, interface{}, bool) {
...
-		s.position = s.at(1)
-		last = s.buckets[s.position]
-	}
-
 	for k, v := range last {
 		delete(last, k)
 		s.buckets[s.at(-1)][k] = v
 		return k, v, false
 	}
+	s.position = s.at(1)
```
Comment (bparees) [inline pkg/image/controller/scheduler.go]: in the interest of (some) clarity, i think i'd be inclined to move this into the "if len(last) == 0" block above and return from within that if-check.  The fact that the for-loop above is really an "if" statement makes it a bit non-obvious what the code flow is here.

**Comment by jim-minter** [comment]
> > return (s.position + inc + len(s.buckets)) % len(s.buckets)
>
> unrelated to your changes but since it hurt my brain, isn't that numerically identical to return (s.position + inc) % len(s.buckets) ?
> isn't (x+y) % y == x % y ?

It's not numerically identical in the case that inc is negative: in this case the return value could be negative and therefore invalid.  I suppose the intended contract here is that abs(inc) <= len(s.buckets).

**Comment by bparees** [comment]
> It's not numerically identical in the case that inc is negative: in this case the return value could be negative and therefore invalid. 

sigh, yes of course.

**Comment by jim-minter** [comment]
> so this does fix the code to behave as the doc implies it was intended to ("If the bucket is empty, we wait for the rate limiter before returning.") but since i haven't studied this like you have, why is it bad to skip through the empty buckets? i'm clear how the buckets, rate limiter, and desired import schedule interact, so it's not obvious to me why treating empty buckets as "an item to process" is critical to it working properly.

Say the scheduler has 4 buckets and the minimum refresh interval is 15 minutes.  I think the idea was that the ratelimiter emits 4 tokens in 15 minutes so that the scheduler gets through all its buckets once in the period.  If the scheduler immediately skips to the next bucket when done without waiting on the ratelimiter, it goes too fast (https://bugzilla.redhat.com/show_bug.cgi?id=1515058).  Worse, if no scheduler bucket is empty, the ratelimiter is never called and the scheduler hot loops (a second rate limiter based on MaxScheduledImageImportsPerMinute meant that this wasn't super obvious).  That's https://bugzilla.redhat.com/show_bug.cgi?id=1543446.

**Comment by jim-minter** [comment]
At the moment I'm looking for backportable correctness fixes because I think this may be appropriate for backport.  Subsequently I'd like to refactor this as I think there are several corner case race conditions and the whole thing is very opaque.

**Comment by bparees** [comment]
>  Worse, if no scheduler bucket is empty, the ratelimiter is never called and the scheduler hot loops

every bucket will eventually be empty since it's moving items out of the bucket as it goes, right?

but thanks for you other explanation, that makes sense (that the ratelimiter config and the number of buckets are configured together.  I didn't want to dig into the code enough to see if that was the case)

**Comment by bparees** [comment]
As for backporting, this doesn't feel like it rises to the level of needing to be backported (especially if we can get it into 3.9) but that would be up to CEE to negotiate w/ the customer who hit the issue.  I'd ask on the bug.

File: pkg/image/controller/scheduler.go (hunk 1 above)
Comment (jim-minter) [reply pkg/image/controller/scheduler.go]: I tried it and don't believe it ends up clearer.  The compiler ends up insisting on a return statement at the end of the function which is actually unreachable.  Or you need to declare k and v outside the range and break.  I've added comments.

**Comment by jim-minter** [comment]
> every bucket will eventually be empty since it's moving items out of the bucket as it goes, right?

That's not relevant - because Add() spreads items across all buckets, any OCP instance with more scheduled imagestreams than buckets will hot loop.  It's in https://bugzilla.redhat.com/show_bug.cgi?id=1543446 and the unit test I've added recreates it.  The net effect is that the backend docker registries are queried constantly to the rate of MaxScheduledImageImportsPerMinute.

File: pkg/image/controller/scheduler_test.go [hunk 2]
```diff
@@ -126,3 +128,102 @@ func TestSchedulerRemove(t *testing.T) {
...
+func (h int64Heap) Less(i, j int) bool  { return h[i] < h[j] }
+func (h int64Heap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
+func (h *int64Heap) Push(x interface{}) { *h = append(*h, x.(int64)) }
+func (h *int64Heap) Pop() interface{} {
+	x := (*h)[len(*h)-1]
+	*h = (*h)[:len(*h)-1]
+	return x
+}
+
+type wallClock struct{}
```
Comment (bparees) [inline pkg/image/controller/scheduler_test.go]: does not appear to be used

**Comment by jim-minter** [comment]
wallClock comment added

**Pull request merged**